Added support for mods to add unique monsters without replacing any data

Andrettin opened the pull request:
At present, if a mod wants to add a unique monster, it has to replace the entire unique_monstdat.tsv file. This pull request allows mods to add unique monsters without replacing any existing data. This makes it easier to create small mods that just add unique monsters, and allows players to have multiple such mods enabled at the same time.

With this change, mods can add unique monsters by including a file called "unique_monstdat_add.tsv". This file has the same structure as the main unique monster data TSV file, and *all* instances of it present are loaded if found, i.e. from all data/mod paths (only for enabled mods, of course).

If the enabled mods change, and a mod-added unique monster is present in a save game, but no longer in the data (i.e. because a mod was disabled), then when the level in which that monster is present is loaded, the monster will no longer be unique. It will otherwise preserve properties granted to it via unique monster spawning however, e.g. increased hit points. Only the Monster::uniqueType field is changed (to UniqueMonsterType::None). A unique monster is considered to still be valid if 1. its index is smaller than the size of the UniqueMonstersData vector, 2. if the monster has the same monster type as the unique monster in data (to account for the case of a mod being disabled and another enabled, and the unique monster index corresponding to a different unique monster) and 3. if the monster's dungeon level is either zero or equal to the loaded level.

Tested:
1. Added a mod with a unique monster defined in "unique_monstdat_add.tsv" -> Verified that the monster is spawned correctly by the game
2. Removed the mod and loaded a saved game with the unique monster -> Verified that the unique monster has become a non-unique monster
3. Re-added the mod and loaded the saved game -> Verified that the unique monster is correctly there

AJenbo commented:
> With this change, mods can add unique monsters by including a file called "unique_monstdat_add.tsv". This file has the same structure as the main unique monster data TSV file, and all instances of it present are loaded if found, i.e. from all data/mod paths (only for enabled mods, of course).

I guess this tries to work around overwriting but it greatly complicates things when some files are loaded in special ways.

> If the enabled mods change, and a mod-added unique monster is present in a save game, but no longer in the data (i.e. because a mod was disabled), then when the level in which that monster is present is loaded, the monster will no longer be unique. It will otherwise preserve properties granted to it via unique monster spawning however, e.g. increased hit points.

Generally my thinking is that save games should be bound to which mods are loaded, we don't have  this mechanism in place yet though.

Andrettin commented:
> A better solution would be to add Lua bindings for adding additional monsters.

That's a great idea! I agree that it would be a better solution. Then that binding function would be run during mod initialization? (via each mod's init.lua)

> Generally my thinking is that save games should be bound to which mods are loaded, we don't have this mechanism in place yet though.

I think that makes sense for big mods, but for small mods I feel that it would make them too cumbersome to use, e.g. requiring a creating a new character just to add a mod which adds some unique monsters. But if it's just about the save file (i.e. the player would still be able to use the same character to start a new run), then I think it's not so bad, indeed.

Maybe a good solution would be to allow loading the games which were saved with a different mod combination, but warn users that their current mod combination is different than the one originally used for creating that save, and that they proceed at their own risk.

Either way, I think it makes sense to handle detected changes in unique monster data gracefully, and try to use the save despite that. Especially since this could be relevant not only because of different mod combinations, but also because e.g. a new version of a mod changed removed a unique monster from before, changing the indices of the remaining unique monsters.

Andrettin commented:
I would then do the following: rework this pull request to use the approach suggested by @AJenbo , using Lua bindings to add new unique monsters (via mod initialization I presume). If there are any suggestions on e.g. the specific signature of the Lua function, let me know :)

AJenbo commented:
> I think that makes sense for big mods, but for small mods I feel that it would make them too cumbersome to use, e.g. requiring a creating a new character just to add a mod which adds some unique monsters. But if it's just about the save file (i.e. the player would still be able to use the same character to start a new run), then I think it's not so bad, indeed.

The problem is poluting the regular game, especially for multiplayer. If one of the players always play in the "all unique monsters" mod then they can rake up good equipment very fast. But it's also tricky because some game content is not as forgiving, if a mod adds just a few new items then that morphs all items when enabling or disabling it.

Andrettin commented:
I reworked the PR now to use Lua bindings to allow defining new unique monsters. This is done like this:

`local monsters = require("devilutionx.monsters")

monsters.addUniqueMonsterData(
	"MT_RFALLSD",  -- type
	"Bonegnasher the Ogre", -- name
	"general", -- trn
	1, -- level
	80, -- maxHp
	"Fat", -- ai
	1, -- intelligence
	6, -- minDamage
	12, -- maxDamage
	"IMMUNE_MAGIC", -- resistance
	"Leashed", -- monsterPack
	0, -- customToHit
	0 -- customArmorClass
)
`

The unique monsters added in this way are stored in their own vector, and then appended to the main unique monsters vector when unique monster data is loaded. I did it this way since otherwise they would get cleared whenever LoadMonsterData() is called (which happens e.g. after mods are loaded).

It wasn't clear to me if I should remove the unique monster validation I had added before, so I kept it. If I should remove it, please let me know, and I'll remove that from the PR.

AJenbo commented on Source/monstdat.cpp:
```diff
@@ -583,6 +597,9 @@ void LoadUniqueMonstDat()
 		});
 	}
+
+	UniqueMonstersData.insert(UniqueMonstersData.end(), AdditionalUniqueMonstersData.begin(), AdditionalUniqueMonstersData.end());
```
I wonder if it would be better to make a Lua trigger for "monster data loaded". That could allow for other things like modifying all loaded monsters.

Andrettin replied:
I think that's a very good suggestion, since it would remove the need for the additional vector and make the code cleaner. I will do that then :)

AJenbo replied:
Thanks for doing this

AJenbo commented on Source/items/validation.cpp:
```diff
@@ -194,4 +194,27 @@ bool IsItemDeltaValid(const TCmdPItem &itemDelta)
 	return IsItemValid(*MyPlayer, item);
 }
 
+bool IsUniqueMonsterValid(const Monster &monster)
```
Monster validation doesn't belong in the `item` validation file

AJenbo commented on Source/loadsave.cpp:
```diff
@@ -736,6 +736,11 @@ void LoadMonster(LoadHelper *file, Monster &monster, MonsterConversionData *mons
+
+	if (monster.isUnique() && !IsUniqueMonsterValid(monster)) {
+		// if the unique monster is no longer valid (e.g. because the loaded mods changed and the unique monsters changed as a consequence), then this monster will cease to be unique, but will otherwise maintain its properties
+		monster.uniqueType = UniqueMonsterType::None;
```
I think i would lean more in the direction of skipping the invalid monster, and we should probably have either an assert or at least a log line here to notify that something unexpected happened in case this starts happening for the wrong reason.

Andrettin commented:
There was an issue with a unit test now failing (it was a false positive, a consequence of loading monster data now triggering a Lua event, and in those tests Lua wasn't initialized). I've fixed that now.

AJenbo commented:
Nice work

Andrettin commented:
> Nice work

Thank you! 😄
